File: src/jfan/fan/sys/ListType.java
//
// Copyright (c) 2006, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   13 Jan 06 (Fri 13th)  Brian Frank  Creation
//
package fan.sys;

import java.io.*;
import java.lang.reflect.*;
import java.util.ArrayList;
import java.util.HashMap;
import fanx.fcode.*;
import fanx.emit.*;

/**
 * ListType is the GenericType for Lists: Foo[] -> V = Foo
 */
public class ListType
  extends GenericType
{

//////////////////////////////////////////////////////////////////////////
// Constructor
//////////////////////////////////////////////////////////////////////////

  ListType(Type v)
  {
    super(Sys.ListType);
    this.v = v;
  }

//////////////////////////////////////////////////////////////////////////
// Type
//////////////////////////////////////////////////////////////////////////

  public final Int hash()
  {
    return signature().hash();
  }

  public final Bool _equals(Object obj)
  {
    if (obj instanceof ListType)
    {
      return v._equals(((ListType)obj).v);
    }
    return Bool.False;
  }

  public final Type base()
  {
    return Sys.ListType;
  }

  public final Str signature()
  {
    if (sig == null)
    {
      sig = Str.make(v.signature().val + "[]");
    }
    return sig;
  }

  public boolean is(Type type)
  {
    if (type instanceof ListType)
    {
      ListType t = (ListType)type;
      return v.is(t.v);
    }
    return super.is(type);
  }

  Map makeParams()
  {
    return new Map(Sys.StrType, Sys.TypeType)
      .set(Str.ascii['V'], v)
      .set(Str.ascii['L'], this).ro();
  }

//////////////////////////////////////////////////////////////////////////
// GenericType
//////////////////////////////////////////////////////////////////////////

  public Type getRawType()
  {
    return Sys.ListType;
  }

  public boolean isGenericParameter()
  {
    return v.isGenericParameter();
  }

  protected Type doParameterize(Type t)
  {
    if (t == Sys.VType) return v;
    if (t == Sys.LType) return this;
    throw new IllegalStateException(t.toString());
  }

//////////////////////////////////////////////////////////////////////////
// Fields
//////////////////////////////////////////////////////////////////////////

  public final Type v;
  private Str sig;

}

File: src/nfan/fan/sys/ListType.cs
//
// Copyright (c) 2006, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   27 Sep 06  Andy Frank  Creation
//

namespace Fan.Sys
{
  /// <summary>
  /// ListType is the GenericType for Lists: Foo[] -> V = Foo
  /// </summary>
  public class ListType : GenericType
  {

  //////////////////////////////////////////////////////////////////////////
  // Constructor
  //////////////////////////////////////////////////////////////////////////

    internal ListType(Type v) : base(Sys.ListType)
    {
      this.m_v = v;
    }

  //////////////////////////////////////////////////////////////////////////
  // Type
  //////////////////////////////////////////////////////////////////////////

    public override Int hash()
    {
      return signature().hash();
    }

    public override Bool _equals(object obj)
    {
      if (obj is ListType)
      {
        return m_v._equals(((ListType)obj).m_v);
      }
      return Bool.False;
    }

    public override Type @base()
    {
      return Sys.ListType;
    }

    public override Str signature()
    {
      if (m_sig == null)
      {
        m_sig = Str.make(m_v.signature().val + "[]");
      }
      return m_sig;
    }

    public override bool @is(Type type)
    {
      if (type is ListType)
      {
        ListType t = (ListType)type;
        return m_v.@is(t.m_v);
      }
      return base.@is(type);
    }

    internal override Map makeParams()
    {
      return new Map(Sys.StrType, Sys.TypeType)
        .set(Str.m_ascii['V'], m_v)
        .set(Str.m_ascii['L'], this).ro();
    }

  //////////////////////////////////////////////////////////////////////////
  // GenericType
  //////////////////////////////////////////////////////////////////////////

    public override Type getRawType()
    {
      return Sys.ListType;
    }

    public override bool isGenericParameter()
    {
      return m_v.isGenericParameter();
    }

    protected override Type doParameterize(Type t)
    {
      if (t == Sys.VType) return m_v;
      if (t == Sys.LType) return this;
      throw new System.InvalidOperationException(t.ToString());
    }

  //////////////////////////////////////////////////////////////////////////
  // Fields
  //////////////////////////////////////////////////////////////////////////

    public readonly Type m_v;
    private Str m_sig;

  }
}